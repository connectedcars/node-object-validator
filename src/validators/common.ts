import { ValidationErrorContext } from '../errors'

export type Schema = {
  [key: string]: Validator
}

export interface Validator {
  schema?: Schema
  validate(value: unknown, context?: ValidationErrorContext): Error | null
}
