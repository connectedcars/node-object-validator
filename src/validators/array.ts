import { ValidatorBase } from '../common'
import { NotArrayError, RequiredError, ValidationErrorContext, WrongLengthError } from '../errors'
import { ValidatorTypes } from '../types'

export function validateArray(
  value: unknown,
  minLength = 0,
  maxLength = Number.MAX_SAFE_INTEGER,
  context?: ValidationErrorContext
): Error[] | null {
  if (!Array.isArray(value)) {
    return [new NotArrayError(`Must be an array (received "${value}")`, context)]
  }
  if ((minLength !== 0 && value.length < minLength) || value.length > maxLength) {
    return [
      new WrongLengthError(
        `Must contain between ${minLength} and ${maxLength} entries (found ${value.length})`,
        context
      )
    ]
  }
  return null
}

export class RequiredArray<T extends ValidatorTypes = ValidatorTypes> extends ValidatorBase {
  public schema: T
  private type: 'RequiredArray' = 'RequiredArray'
  private minLength: number
  private maxLength: number

  public constructor(schema: T, minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    super()
    this.schema = schema
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] | null {
    if (value == null) {
      return [new RequiredError(`Is required`, context)]
    }
    return validateArray(value, this.minLength, this.maxLength, context)
  }
}

export class OptionalArray<T extends ValidatorTypes = ValidatorTypes> extends ValidatorBase {
  public schema: T
  private type: 'OptionalArray' = 'OptionalArray'
  private minLength: number
  private maxLength: number

  public constructor(schema: T, minLength = 0, maxLength = Number.MAX_SAFE_INTEGER) {
    super()
    this.schema = schema
    this.minLength = minLength
    this.maxLength = maxLength
  }

  public validate(value: unknown, context?: ValidationErrorContext): Error[] | null {
    if (value == null) {
      return null
    }
    return validateArray(value, this.minLength, this.maxLength, context)
  }
}

export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength: number,
  maxLength: number,
  required?: false
): OptionalArray<T>
export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength: number,
  maxLength: number,
  required: true
): RequiredArray<T>
export function TypedArray<T extends ValidatorTypes = ValidatorTypes>(
  schema: T,
  minLength = 0,
  maxLength = Number.MAX_SAFE_INTEGER,
  required = false
): OptionalArray<T> | RequiredArray<T> {
  return required
    ? new RequiredArray<T>(schema, minLength, maxLength)
    : new OptionalArray<T>(schema, minLength, maxLength)
}
