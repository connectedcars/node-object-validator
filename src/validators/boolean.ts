import { CodeGenResult, ValidatorBase, ValidatorBaseOptions, ValidatorExportOptions, ValidatorOptions } from '../common'
import { NotBooleanFail, RequiredFail, ValidationFailure } from '../errors'

export function isBoolean(value: unknown, context?: string): value is boolean {
  const errors = validateBoolean(value, context)
  if (errors.length === 0) {
    return true
  }
  return false
}

export function validateBoolean(value: unknown, context?: string): ValidationFailure[] {
  if (typeof value !== 'boolean') {
    return [new NotBooleanFail(`Must be an boolean`, value, context)]
  }
  return []
}

export abstract class BooleanValidator<O = never> extends ValidatorBase<boolean | O> {
  public constructor(options?: ValidatorBaseOptions) {
    super(options)
    if (options?.optimize !== false) {
      this.optimize()
    }
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: string,
    earlyFail?: boolean
  ): CodeGenResult {
    const contextStr = context ? `, \`${context}\`` : ', context'
    const localValueRef = `value${id()}`
    const declarations: string[] = []
    // prettier-ignore
    const code: string[] = [
      `const ${localValueRef} = ${valueRef}`,
      ...this.nullCheckWrap([
      `  if (typeof ${localValueRef} !== 'boolean') {`,
      `    errors.push(new NotBooleanFail(\`Must be an boolean\`, ${localValueRef}${contextStr}))`,
      `  }`,
      ], localValueRef, contextStr),
      ...(earlyFail ? [
      `if (errors.length > 0) {`,
      `  return errors`,
      `}`] : []),
    ]
    return [
      {
        NotBooleanFail: NotBooleanFail,
        RequiredFail: RequiredFail
      },
      declarations,
      code
    ]
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString()
    }
  }

  protected validateValue(value: unknown, context?: string): ValidationFailure[] {
    return validateBoolean(value, context)
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        let typeStr = `boolean`

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {
        const isOption = !this.required || this.nullable
        return isOption ? `Option<bool>` : 'bool'
      }
      default: {
        throw new Error(`Language: '{}' unknown`)
      }
    }
  }

  private constructorString(): string {
    return `new ${this.constructor.name}(${this.optionsString})`
  }
}

export class RequiredBoolean extends BooleanValidator {
  public constructor(options?: ValidatorOptions) {
    super({ ...options })
  }
}

export class OptionalBoolean extends BooleanValidator<undefined> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false })
  }
}

export class NullableBoolean extends BooleanValidator<null> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, nullable: true })
  }
}

export class OptionalNullableBoolean extends BooleanValidator<undefined | null> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false, nullable: true })
  }
}
