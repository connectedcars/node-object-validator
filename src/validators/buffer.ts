import { ValidatorBase, ValidatorBaseOptions, ValidatorExportOptions, ValidatorOptions } from '../common'
import { NotBufferFail, ValidationFailure } from '../errors'

export function isBuffer(value: unknown, context?: string): value is Buffer {
  const errors = validateBuffer(value, context)
  if (errors.length === 0) {
    return true
  }
  return false
}

export function validateBuffer(value: unknown, context?: string): ValidationFailure[] {
  if (Buffer.isBuffer(value)) {
    return []
  }
  return [new NotBufferFail(`Must be an Buffer`, value, context)]
}

export abstract class BufferValidator<O = never> extends ValidatorBase<Buffer | O> {
  public constructor(options?: ValidatorBaseOptions) {
    super(options)
    if (options?.optimize !== false) {
      this.optimize()
    }
  }

  public toString(options?: ValidatorExportOptions): string {
    if (options?.types === true) {
      return this.typeString(options)
    } else {
      return this.constructorString()
    }
  }

  protected validateValue(value: unknown, context?: string): ValidationFailure[] {
    return validateBuffer(value, context)
  }

  private typeString(options?: ValidatorExportOptions): string {
    const language = options?.language ?? 'typescript'
    switch (language) {
      case 'typescript': {
        let typeStr = `Buffer`

        if (this.required === false) {
          typeStr += ` | undefined`
        }
        if (this.nullable === true) {
          typeStr += ` | null`
        }

        return typeStr
      }
      case 'rust': {
        throw new Error('TODO viktor')
      }
      default: {
        throw new Error(`Language: '${options?.language}' unknown`)
      }
    }
  }

  private constructorString(): string {
    return `new ${this.constructor.name}(${this.optionsString})`
  }
}

export class RequiredBuffer extends BufferValidator {
  public constructor(options?: ValidatorOptions) {
    super({ ...options })
  }
}

export class OptionalBuffer extends BufferValidator<undefined> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false })
  }
}

export class NullableBuffer extends BufferValidator<null> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, nullable: true })
  }
}

export class OptionalNullableBuffer extends BufferValidator<undefined | null> {
  public constructor(options?: ValidatorOptions) {
    super({ ...options, required: false, nullable: true })
  }
}
