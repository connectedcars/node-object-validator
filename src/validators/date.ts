import { CodeGenResult, ValidatorBase, ValidatorOptions } from '../common'
import { NotDateFail, RequiredFail, ValidationErrorContext, ValidationFailure } from '../errors'

export function validateDate(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
  if (!(value instanceof Date)) {
    return [new NotDateFail(`Must be a Date object`, context)]
  }
  return []
}

export class DateValidator<O = never> extends ValidatorBase<Date | O> {
  private required: boolean

  public constructor(options?: ValidatorOptions, required = true) {
    super()
    this.required = required
    if (options?.optimize) {
      this.validate = this.optimize()
    }
  }

  public validate(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
    if (value == null) {
      return this.required ? [new RequiredFail(`Is required`, context)] : []
    }
    return validateDate(value, context)
  }

  public codeGen(
    valueRef: string,
    validatorRef: string,
    id = () => {
      return this.codeGenId++
    },
    context?: ValidationErrorContext
  ): CodeGenResult {
    const contextStr = context ? `, { key: \`${context.key}\` }` : ', context'
    const localValueRef = `value${id()}`
    const declarations: string[] = []
    // prettier-ignore
    const code: string[] = [
      `const ${localValueRef} = ${valueRef}`,
      `if (${localValueRef} != null) {`,
      `  if (!(${localValueRef} instanceof Date)) {`,
      `    errors.push(new NotDateFail(\`Must be a Date object\`${contextStr}))`,
      `  }`,
      ...(this.required ? [
        `} else {`,
        `  errors.push(new RequiredError(\`Is required\`${contextStr}))`] : []),
        '}'
    ]
    return [
      {
        NotDateFail: NotDateFail,
        RequiredError: RequiredFail
      },
      declarations,
      code
    ]
  }
}

export class RequiredDate extends DateValidator {
  private validatorType: 'RequiredDate' = 'RequiredDate'
  public constructor(options?: ValidatorOptions) {
    super(options)
  }
}

export class OptionalDate extends DateValidator<null | undefined> {
  private validatorType: 'OptionalDate' = 'OptionalDate'
  public constructor(options?: ValidatorOptions) {
    super(options, false)
  }
}

export function DateObject(required: false): OptionalDate
export function DateObject(required?: true): RequiredDate
export function DateObject(required = true): OptionalDate | RequiredDate {
  return required ? new RequiredDate() : new OptionalDate()
}
