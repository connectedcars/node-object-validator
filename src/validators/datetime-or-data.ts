// XXX: This is a temporary solution until we support unions

import { validateDate, validateDateTime } from '..'
import { isValidType, ValidatorBase, ValidatorOptions } from '../common'
import { NotDatetimeOrDateFail, RequiredFail, ValidationErrorContext, ValidationFailure } from '../errors'

export function validateDateTimeOrDate(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
  const dateError = validateDate(value, context)
  if (!isValidType<Date>(value, dateError)) {
    const dateTimeError = validateDateTime(value, context)
    if (!isValidType<string>(value, dateTimeError)) {
      return [new NotDatetimeOrDateFail(`Must be a Date or a formatted as an RFC 3339 timestamp (received "${value}")`)]
    }
    return [] //valid datetime
  }
  return [] //valid Date
}

export class DateTimeOrDateValidator<O = never> extends ValidatorBase<string | O> {
  public required: boolean

  public constructor(options?: ValidatorOptions, required = true) {
    super()
    this.required = required
    if (options?.optimize) {
      //TODO: fix optimise
      //this.validate = this.optimize()
    }
  }

  public validate(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
    if (value == null) {
      return this.required ? [new RequiredFail(`Is required`, context)] : []
    }
    return validateDateTimeOrDate(value, context)
  }
}

export class RequiredDateTimeOrDate extends DateTimeOrDateValidator {
  private validatorType: 'RequiredDateTimeOrDateValidator' = 'RequiredDateTimeOrDateValidator'
  public constructor(options?: ValidatorOptions) {
    super(options)
  }
}

export class OptionalDateTimeOrDate extends DateTimeOrDateValidator<undefined | null> {
  private validatorType: 'OptionalDateTimeOrDateValidator' = 'OptionalDateTimeOrDateValidator'
  public constructor(options?: ValidatorOptions) {
    super(options, false)
  }
}

export function DateTimeOrDate(required: false): OptionalDateTimeOrDate
export function DateTimeOrDate(required?: true): RequiredDateTimeOrDate
export function DateTimeOrDate(required = true): OptionalDateTimeOrDate | RequiredDateTimeOrDate {
  return required ? new RequiredDateTimeOrDate() : new OptionalDateTimeOrDate()
}
