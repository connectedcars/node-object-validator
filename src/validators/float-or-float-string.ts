// XXX: This is a temporary solution until we support unions

import { validateFloat, validateFloatString } from '..'
import { isValidType, ValidatorBase, ValidatorOptions } from '../common'
import { NotFloatOrFloatStringFail, RequiredFail, ValidationErrorContext, ValidationFailure } from '../errors'

export function validateFloatOrFloatString(
  value: unknown,
  min: number,
  max: number,
  context?: ValidationErrorContext
): ValidationFailure[] {
  const floatError = validateFloat(value, min, max, context)
  if (!isValidType<number>(value, floatError)) {
    const floatStringError = validateFloatString(value, min, max, context)
    if (!isValidType<string>(value, floatStringError)) {
      if (min === Number.MIN_SAFE_INTEGER && max === Number.MAX_SAFE_INTEGER) {
        return [new NotFloatOrFloatStringFail(`Must be a float or a string formatted float (received "${value}")`)]
      }
      if (min === Number.MIN_SAFE_INTEGER && max !== Number.MAX_SAFE_INTEGER) {
        return [
          new NotFloatOrFloatStringFail(
            `Must be a float or a string formatted float smaller than ${max} (received "${value}")`
          )
        ]
      }
      if (min !== Number.MIN_SAFE_INTEGER && max === Number.MAX_SAFE_INTEGER) {
        return [
          new NotFloatOrFloatStringFail(
            `Must be a float or a string formatted float larger than ${min} (received "${value}")`
          )
        ]
      }
      if (min !== Number.MIN_SAFE_INTEGER && max !== Number.MAX_SAFE_INTEGER) {
        return [
          new NotFloatOrFloatStringFail(
            `Must be a float or a string formatted float between ${min} and ${max} (received "${value}")`
          )
        ]
      }
    }
    return [] //valid floatstring
  }
  return [] //valid float
}

export class FloatOrFloatStringValidator<O = never> extends ValidatorBase<string | O> {
  private min: number
  private max: number

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(options)
    this.min = min
    this.max = max
    if (options?.optimize) {
      this.optimize()
    }
  }

  public validate(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
    if (value == null) {
      return this.required ? [new RequiredFail(`Is required`, context)] : []
    }
    return validateFloatOrFloatString(value, this.min, this.max, context)
  }
}

export class RequiredFloatOrFloatString extends FloatOrFloatStringValidator {
  public constructor(min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(min, max, { ...options, required: true })
  }
}

export class OptionalFloatOrFloatString extends FloatOrFloatStringValidator<undefined | null> {
  public constructor(min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(min, max, { ...options, required: false })
  }
}
