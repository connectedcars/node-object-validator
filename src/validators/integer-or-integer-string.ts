// XXX: This is a temporary solution until we support unions

import { validateInteger, validateIntegerString } from '..'
import { isValidType, ValidatorBase, ValidatorOptions } from '../common'
import { NotIntegerOrIntegerStringFail, RequiredFail, ValidationErrorContext, ValidationFailure } from '../errors'

export function validateIntegerOrIntegerString(
  value: unknown,
  min: number,
  max: number,
  context?: ValidationErrorContext
): ValidationFailure[] {
  const integerError = validateInteger(value, min, max, context)
  if (!isValidType<number>(value, integerError)) {
    const integerStringError = validateIntegerString(value, min, max, context)
    if (!isValidType<string>(value, integerStringError)) {
      if (min === Number.MIN_SAFE_INTEGER && max === Number.MAX_SAFE_INTEGER) {
        return [
          new NotIntegerOrIntegerStringFail(`Must be a integer or a string formatted integer (received "${value}")`)
        ]
      }
      if (min === Number.MIN_SAFE_INTEGER && max !== Number.MAX_SAFE_INTEGER) {
        return [
          new NotIntegerOrIntegerStringFail(
            `Must be a integer or a string formatted integer smaller than ${max} (received "${value}")`
          )
        ]
      }
      if (min !== Number.MIN_SAFE_INTEGER && max === Number.MAX_SAFE_INTEGER) {
        return [
          new NotIntegerOrIntegerStringFail(
            `Must be a integer or a string formatted integer larger than ${min} (received "${value}")`
          )
        ]
      }
      if (min !== Number.MIN_SAFE_INTEGER && max !== Number.MAX_SAFE_INTEGER) {
        return [
          new NotIntegerOrIntegerStringFail(
            `Must be a integer or a string formatted integer between ${min} and ${max} (received "${value}")`
          )
        ]
      }
    }
    return [] //valid integerstring
  }
  return [] //valid integer
}

export class IntegerOrIntegerStringValidator<O = never> extends ValidatorBase<string | O> {
  private min: number
  private max: number
  private required: boolean

  public constructor(min = 0, max = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super()
    this.min = min
    this.max = max
    const mergedOptions = { required: true, optimize: false, ...options }
    this.required = mergedOptions.required
  }

  public validate(value: unknown, context?: ValidationErrorContext): ValidationFailure[] {
    if (value == null) {
      return this.required ? [new RequiredFail(`Is required`, context)] : []
    }
    return validateIntegerOrIntegerString(value, this.min, this.max, context)
  }
}

export class RequiredIntegerOrIntegerString extends IntegerOrIntegerStringValidator {
  private validatorType: 'RequiredIntegerOrIntegerString' = 'RequiredIntegerOrIntegerString'

  public constructor(min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(min, max, { ...options, required: true })
  }
}

export class OptionalIntegerOrIntegerString extends IntegerOrIntegerStringValidator<undefined | null> {
  private validatorType: 'OptionalIntegerString' = 'OptionalIntegerString'

  public constructor(min = Number.MIN_SAFE_INTEGER, max = Number.MAX_SAFE_INTEGER, options?: ValidatorOptions) {
    super(min, max, { ...options, required: false })
  }
}
