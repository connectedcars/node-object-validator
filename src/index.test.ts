import { ObjectValidator } from './index'
import { DateTime, ExactString, Float, Integer, StringValue, TypedArray, TypedObject } from './index'
import { FloatOrFloatString } from './validators/float-or-float-string'
import { IntegerOrIntegerString } from './validators/integer-or-integer-string'

describe.each([false, true])('Shorthand validation of complex objects (optimize: %s)', optimize => {
  const validator = new ObjectValidator(
    {
      type: ExactString('gps_odometer_km'),
      unitId: StringValue(1, 32),
      recordedAt: DateTime(),
      tripId: Integer(0, 4294967295),
      value: Integer(0, 999999),
      position: TypedObject({
        latitude: Float(-90, 90),
        longitude: Float(-180, 180),
        accuracy: Integer(0, 20),
        extra: TypedObject({
          tag: StringValue(1, 50),
          tagversion: IntegerOrIntegerString(1, 50),
          tagDepth: FloatOrFloatString(0, 42.4)
        })
      }),
      optionalPosition: TypedObject(
        {
          latitude: Float(-90, 90),
          longitude: Float(-180, 180),
          accuracy: Integer(0, 20)
        },
        false
      ),
      positions: TypedArray(
        TypedObject({
          latitude: Float(-90, 90),
          longitude: Float(-180, 180),
          accuracy: Integer(0, 20)
        }),
        2,
        10
      )
    },
    { optimize }
  )

  it('validates successfully', () => {
    expect(
      validator.validate({
        type: 'gps_odometer_km',
        unitId: '1234567',
        recordedAt: '2018-08-06T13:37:00Z',
        tripId: 1337,
        value: 500,
        position: {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18,
          extra: {
            tag: 'yes',
            tagversion: 4,
            tagDepth: '27.89'
          }
        },
        positions: [
          {
            latitude: 55.332131,
            longitude: 12.54454,
            accuracy: 18
          },
          {
            latitude: 55.332131,
            longitude: 12.54454,
            accuracy: 18
          }
        ]
      })
    ).toEqual([])
  })

  it('fails due to empty object', () => {
    const errors = validator.validate({})
    expect(errors.map(e => e.toString())).toMatchSnapshot()
  })

  it('fails due to invalid position', () => {
    const errors = validator.validate({
      type: 'gps_odometer_km',
      unitId: '1234567',
      recordedAt: '2018-08-06T13:37:00Z',
      tripId: 1337,
      value: 500,
      position: {
        latitude: 55.332131,
        longitude: -181,
        accuracy: 18,
        extra: {
          tag: 'bogus',
          tagversion: '4',
          tagDepth: 27
        }
      },
      positions: [
        {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18
        },
        {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18
        }
      ]
    })
    expect(errors.map(e => e.toString())).toEqual([
      `OutOfRangeFail: Field 'position['longitude']' must be between -180 and 180 (received "-181")`
    ])
  })

  it('fails due to invalid double nested object', () => {
    const errors = validator.validate({
      type: 'gps_odometer_km',
      unitId: '1234567',
      recordedAt: '2018-08-06T13:37:00Z',
      tripId: 1337,
      value: 500,
      position: {
        latitude: 55.332131,
        longitude: 12.54454,
        accuracy: 18,
        extra: {
          tag: '',
          tagversion: 32,
          tagDepth: 3.1416
        }
      },
      positions: [
        {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18
        },
        {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18
        }
      ]
    })
    expect(errors.map(e => e.toString())).toEqual([
      `WrongLengthFail: Field 'position['extra']['tag']' must contain between 1 and 50 characters (received "")`
    ])
  })

  it('fails due to missing positions', () => {
    const errors = validator.validate({
      type: 'gps_odometer_km',
      unitId: '1234567',
      recordedAt: '2018-08-06T13:37:00Z',
      tripId: 1337,
      value: 500,
      position: {
        latitude: 55.332131,
        longitude: 12.54454,
        accuracy: 18,
        extra: {
          tag: 'test',
          tagversion: 32,
          tagDepth: 3.1416
        }
      }
    })
    expect(errors.map(e => e.toString())).toEqual([`RequiredFail: Field 'positions' is required`])
  })

  it('fails due to too few positions', () => {
    const errors = validator.validate({
      type: 'gps_odometer_km',
      unitId: '1234567',
      recordedAt: '2018-08-06T13:37:00Z',
      tripId: 1337,
      value: 500,
      position: {
        latitude: 55.332131,
        longitude: 12.54454,
        accuracy: 18,
        extra: {
          tag: 'test',
          tagversion: 32,
          tagDepth: 3.1416
        }
      },
      positions: [
        {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18
        }
      ]
    })
    expect(errors.map(e => e.toString())).toEqual([
      `WrongLengthFail: Field 'positions' must contain between 2 and 10 entries (found 1)`
    ])
  })

  it('fails due to invalid positions', () => {
    const errors = validator.validate({
      type: 'gps_odometer_km',
      unitId: '1234567',
      recordedAt: '2018-08-06T13:37:00Z',
      tripId: 1337,
      value: 500,
      position: {
        latitude: 55.332131,
        longitude: 12.54454,
        accuracy: 18,
        extra: {
          tag: 'test',
          tagversion: 32,
          tagDepth: 3.1416
        }
      },
      positions: [
        {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 18
        },
        {
          latitude: 55.332131,
          longitude: 12.54454,
          accuracy: 21
        }
      ]
    })
    expect(errors.map(e => e.toString())).toEqual([
      `OutOfRangeFail: Field 'positions[1]['accuracy']' must be between 0 and 20 (received "21")`
    ])
  })
})
