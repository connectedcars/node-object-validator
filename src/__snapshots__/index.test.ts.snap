// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shorthand validation of complex objects (optimize: false) fails due to empty object 1`] = `
Array [
  "RequiredFail: Field 'type' is required",
  "RequiredFail: Field 'unitId' is required",
  "RequiredFail: Field 'recordedAt' is required",
  "RequiredFail: Field 'tripId' is required",
  "RequiredFail: Field 'value' is required",
  "RequiredFail: Field 'position' is required",
  "RequiredFail: Field 'positions' is required",
]
`;

exports[`Shorthand validation of complex objects (optimize: true) fails due to empty object 1`] = `
Array [
  "RequiredFail: Field 'type' is required",
  "RequiredFail: Field 'unitId' is required",
  "RequiredFail: Field 'recordedAt' is required",
  "RequiredFail: Field 'tripId' is required",
  "RequiredFail: Field 'value' is required",
  "RequiredFail: Field 'position' is required",
  "RequiredFail: Field 'positions' is required",
]
`;
